### PR TITLE
[ENH] use nilearn `logger.log` function instead of `print` or `sys.stderr.write`

### LINKED ISSUE #4443: Use nilearn's logger instread of print statement
There are quite a few print statement in the nilearn codebase that could be replaced by Nilearn's logger.

This would avoid a whole bunch of if blocks and would make it easier to style how messages are printed in the whole code base.

<details>
<summary>All prints</summary>
<pre>
$ grep -nr "print(" nilearn/**/*.py
nilearn/datasets/atlas.py:1857:        print(f"Separating talairach atlas levels: {_TALAIRACH_LEVELS}")
nilearn/datasets/atlas.py:1862:            print(level_name)
nilearn/datasets/func.py:795:                print(f"Skipping path '{path}'...")
nilearn/datasets/func.py:2699:    print("Data absent, downloading...")
nilearn/datasets/func.py:2709:        print("Archive corrupted, trying to download it again.")
nilearn/datasets/func.py:2742:            print(f"{file_name} missing from filelist!")
nilearn/datasets/func.py:2887:        print(
nilearn/datasets/func.py:2902:        print(f"Missing session file: {sess_trials}")
nilearn/datasets/func.py:2912:        print("Missing structural image.")
nilearn/datasets/func.py:2959:    print("Data absent, downloading...")
nilearn/datasets/func.py:2979:            print("Archive corrupted, trying to download it again.")
nilearn/datasets/func.py:3106:                print(f"Missing functional scan for session {int(run)}.")
nilearn/datasets/func.py:3114:                print(f"Missing run file: {sess_dmtx}")
nilearn/datasets/func.py:3122:            print("Missing mask image.")
nilearn/datasets/func.py:3137:    print("Data absent, downloading...")
nilearn/datasets/func.py:3145:        print("Archive corrupted, trying to download it again.")
nilearn/datasets/neurovault.py:914:    print(message)
nilearn/datasets/neurovault.py:1528:        print("Resampling...")
nilearn/datasets/_utils.py:164:            print("Warning: total size could not be determined.")
nilearn/datasets/_utils.py:166:                print(f"Full stack trace: {e}")
nilearn/datasets/_utils.py:233:        print(f"Dataset search paths: {paths}")
nilearn/datasets/_utils.py:244:                print(f"\nDataset found in {path}\n")
nilearn/datasets/_utils.py:260:                    print(f"\nDataset created in {path}\n")
nilearn/datasets/_utils.py:286:                print(f"\nAdded README.md to {d}\n")
nilearn/datasets/_utils.py:375:            print(f"Error uncompressing file: {e}")
nilearn/datasets/_utils.py:582:            print(f"Downloading data from {displayed_url} ...")
nilearn/datasets/_utils.py:613:                    print("Resuming failed, try to download the whole file.")
nilearn/decoding/fista.py:193:            print(
nilearn/decoding/fista.py:201:                print(f"\tConverged (|dE| < {tol:g})")
nilearn/decoding/fista.py:230:                print("decreased dgap_tol")
nilearn/decoding/fista.py:242:                print("Monotonous FISTA: Switching to ISTA")
nilearn/decoding/_proximal_operators.py:250:                    print(
nilearn/decoding/_proximal_operators.py:270:                    print(
nilearn/decoding/space_net.py:245:                    print(
nilearn/decoding/space_net.py:254:            print(f"Test score: {score:.8f}")
nilearn/decoding/space_net.py:1007:            print(
nilearn/decomposition/_base.py:443:            print(f"[{self.__class__.__name__}] Loading data")
nilearn/decomposition/dict_learning.py:285:            print("[DictLearning] Learning initial components")
nilearn/decomposition/dict_learning.py:291:            print("[DictLearning] Computing initial loadings")
nilearn/decomposition/dict_learning.py:299:            print("[DictLearning] Learning dictionary")
nilearn/glm/contrasts.py:239:            print("Automatically converted multi-dimensional t to F contrast")
nilearn/glm/first_level/experimental_paradigm.py:151:        print(
nilearn/glm/first_level/first_level.py:1613:    print(
nilearn/image/image.py:694:     >>> print(joint_mni_image.shape)
nilearn/image/image.py:700:     >>> print(single_mni_image.shape)
nilearn/image/image.py:706:     >>> print(five_mni_images.shape)
nilearn/image/image.py:711:     >>> print(first_three_images.shape)
nilearn/image/image.py:1594:            print(f"Concatenating {index + 1}: {nii_str}")
nilearn/interfaces/tests/test_bids.py:208:    print(new_json_file)
nilearn/maskers/base_masker.py:69:        print(
nilearn/maskers/base_masker.py:97:            print(f"[{class_name}] Resampling images")
nilearn/maskers/base_masker.py:116:            print(f"[{class_name}] Smoothing images")
nilearn/maskers/base_masker.py:125:        print(f"[{class_name}] Extracting region signals")
nilearn/maskers/base_masker.py:140:        print(f"[{class_name}] Cleaning extracted signals")
nilearn/maskers/multi_nifti_masker.py:202:            print(
nilearn/maskers/multi_nifti_masker.py:210:                print("[{self.__class__.__name__}.fit] Computing mask")
nilearn/maskers/multi_nifti_masker.py:264:            print(f"[{self.__class__.__name__}.transform] Resampling mask")
nilearn/maskers/nifti_labels_masker.py:699:                    print("Resampling mask")
nilearn/maskers/nifti_labels_masker.py:792:            print("Resampling labels")
nilearn/maskers/nifti_maps_masker.py:414:                print("Resampling maps")
nilearn/maskers/nifti_maps_masker.py:427:                print("Resampling mask")
nilearn/maskers/nifti_maps_masker.py:544:                    print("Resampling maps")
nilearn/maskers/nifti_maps_masker.py:561:                    print("Resampling mask")
nilearn/maskers/nifti_masker.py:431:            print(
nilearn/maskers/nifti_masker.py:447:                print(f"[{self.__class__.__name__}.fit] Computing the mask")
nilearn/maskers/nifti_masker.py:470:            print(f"[{self.__class__.__name__}.fit] Resampling mask")
nilearn/maskers/nifti_masker.py:493:            print(f"[{self.__class__.__name__}.fit] Finished fit")
nilearn/masking.py:295:        print("EPI mask computation")
nilearn/masking.py:484:        print("Background mask computation")
nilearn/masking.py:637:        print("Template", mask_type, "mask computation")
nilearn/plotting/img_plotting.py:1944:                print(f"Coercing atlas_values to {label_dtype}")
nilearn/regions/parcellations.py:363:            print(f"[{self.__class__.__name__}] computing {self.method}")
nilearn/regions/rena_clustering.py:419:            print(
nilearn/_utils/extmath.py:66:            print(f"matrix not symmetric to {decimal:d} decimals")
nilearn/_utils/extmath.py:71:        print(f"matrix has a negative eigenvalue: {eigvalsh.min():.3f}")
nilearn/_utils/param_validation.py:155:        print(
nilearn/_utils/param_validation.py:158:        print(
nilearn/_utils/param_validation.py:163:        print(
nilearn/_utils/param_validation.py:166:        print(
</pre>
</details>

Nilearn logger is here:

https://github.com/nilearn/nilearn/blob/8602551179dc4112745cd2dc15ed3910b607c275/nilearn/_utils/logger.py#L12

So don't do:

```python
verbose = 1
if verbose:
   print("foo")
```

but do

```python
from nilearn._utils import logger

versbose = 1

logger.log("foo", versbose = versbose)
```

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes #4443 

Changes proposed in this pull request:

- use nilearn logger instead of "simple" `print`

### TODO

- [x] use the stack level parameter of the nilearn logger to check that messages are not displayed as being sent from a private function / method

### CODE REVIEW

File: nilearn/maskers/base_masker.py [hunk 1]
```diff
@@ -55,21 +54,16 @@ def _filter_and_extract(
     """
     if memory is None:
         memory = Memory(location=None)
-    # Since the calling class can be any *Nifti*Masker, we look for exact type
-    if verbose > 0:
-        class_name = enclosing_scope_name(stack_level=10)
```
Comment (Remi-Gau) [inline nilearn/maskers/base_masker.py:60]: Only place in the codebase where `enclosing_scope_name` is used: consider removing?

Or integrate into the logger module?

TODO: check that the output printed by the masker loggers still make sense after this change.

File: nilearn/datasets/neurovault.py [hunk 2]
```diff
@@ -909,6 +914,8 @@ def _print_if(message, level, threshold_level, with_traceback=False):
         if `message` is printed, also print the last traceback.
 
     """
+    # TODO replace by the nileanr logger,
+    # that does almost exactly the same thing!
     if level > threshold_level:
```
Comment (Remi-Gau) [inline nilearn/datasets/neurovault.py:919]: Possible to remove an entire private function from this and only rely on the nilearn logger that does the same thing.

File: nilearn/interfaces/tests/test_bids.py [hunk 3]
```diff
@@ -205,7 +205,6 @@ def test_get_bids_files_inheritance_principle_sub_folder(tmp_path, json_file):
         metadata={"RepetitionTime": 1.5},
         json_file=json_file,
     )
-    print(new_json_file)
```
Comment (Remi-Gau) [inline nilearn/interfaces/tests/test_bids.py:208]: removing as this looks more a left over of a "print debugging" session (my fault)

File: nilearn/decoding/space_net.py [hunk 4]
```diff
@@ -240,18 +239,15 @@ def __call__(self, variables):
...
-                    )
-                else:
-                    sys.stderr.write(".")
+                message = (
+                    f"Early stopping. \n" f"Test score: {score:.8f} {40 * '-'}"
+                )
+            logger.log(message)
             return True
 
-        if self.verbose > 1:
```
Comment (Remi-Gau) [inline nilearn/decoding/space_net.py:253]: TODO: some part of the code have a different default value for verbosity, make sure the printing behavior is unchanged.

File: nilearn/datasets/func.py [hunk 5]
```diff
@@ -16,17 +16,12 @@
 import numpy as np
 import pandas as pd
 from scipy.io import loadmat
-
-try:
-    from scipy.io.matlab import MatReadError
-except ImportError:  # SciPy < 1.8
-    from scipy.io.matlab.miobase import MatReadError
```
Comment (Remi-Gau) [inline nilearn/datasets/func.py:23]: AFAICT this is not needed anymore as we only support scipy >= 1.8

File: nilearn/_utils/param_validation.py [hunk 6]
```diff
@@ -151,21 +153,24 @@ def adjust_screening_percentile(screening_percentile, mask_img, verbose=0):
...
-            f"= {MNI152_BRAIN_VOLUME:g}mm^3 "
-            f"= {MNI152_BRAIN_VOLUME / 1.0e3:g}cm^3"
-        )
-        print(
-            f"Original screening-percentile: {original_screening_percentile:g}"
-        )
-        print(
-            f"Volume-corrected screening-percentile: {screening_percentile:g}"
-        )
+    logger.log(
```
Comment (Remi-Gau) [inline nilearn/_utils/param_validation.py:156]: TODO: need to add msg_lvl

File: nilearn/decoding/space_net.py [hunk 7]
```diff
@@ -240,18 +239,15 @@ def __call__(self, variables):
...
+            message = "."
             if self.verbose:
-                if self.verbose > 1:
-                    print(
-                        "Early stopping. "
-                        f"Test score: {score:.8f} {40 * '-'}"
-                    )
-                else:
-                    sys.stderr.write(".")
+                message = (
```
Comment (Remi-Gau) [inline nilearn/decoding/space_net.py:244]: TODO: need to add msg_lvl

File: nilearn/decoding/space_net.py (hunk 4 above)
Comment (Remi-Gau) [reply nilearn/decoding/space_net.py:253]: TODO: need to add msg_lvl

File: nilearn/maskers/multi_nifti_masker.py [hunk 8]
```diff
@@ -198,16 +199,15 @@ def fit(self, imgs=None, y=None):
 
         """
         # Load data (if filenames are given, load them)
-        if self.verbose > 0:
-            print(
-                f"[{self.__class__.__name__}.fit] Loading data from "
-                f"{_utils._repr_niimgs(imgs, shorten=False)}."
-            )
+        logger.log(
+            f"[{self.__class__.__name__}.fit] Loading data from "
```
Comment (Remi-Gau) [inline nilearn/maskers/multi_nifti_masker.py:203]:
```suggestion
            f"Loading data from "
```

File: nilearn/_utils/class_inspect.py [hunk 9]
```diff
@@ -46,44 +40,3 @@ class 'cls' and returns the value for these parameters in object
             params[param_name] = getattr(instance, param_name)
 
     return params
-
-
-def enclosing_scope_name(ensure_estimator=True, stack_level=2):
```
Comment (Remi-Gau) [inline nilearn/_utils/class_inspect.py:51]: the nilearn logger seems to be already doing this

**Review by bthirion — COMMENTED** [review]
few lines seem to be untested. LGTM otherwise.

**Comment by Remi-Gau** [comment]
> few lines seem to be untested. LGTM otherwise.

am going to try to cover some but I suspect those lines were not hit before either as they are just print statements for some error cases

**Comment by bthirion** [comment]
Thr PR LGTM but there remain some codecov complaints.

**Comment by Remi-Gau** [comment]
OK the most obvious way to improve coverage would require tweaking the way some of the fetchers are tested so this is beyond the scope of this PR.

Will merge as is before this thing grows even more.

**Pull request merged**